### PR TITLE
RES: uniform handling of associated items

### PR DESCRIPTION
**Pull request by matklad**
r? @vlad20012

### CODE REVIEW

**Comment by vlad20012** [comment]
docs again?

**Comment by matklad** [comment]
Yeah, seems like this is a genuinely non-deterministic test :( I'd like to see if it fails deterministically on travis.

**Comment by matklad** [comment]
Hm, nope, that's an actual resolve failure, which was not caught by our resolve test:

```
    fun `test foo`() = checkByCode("""
        pub trait Iter {
            type Item;
                //X

            fn scan<St, B, F>(self, initial_state: St, f: F) -> Scan<Self, St, F>
                where Self: Sized, F: FnMut(&mut St, Self::Item) -> Option<B> { unimplemented!() }
        }                                                 //^
    """)
```

**Comment by notriddle** [comment]
Protip: if you're trying to find out if a test is deterministic or not, you probably want to use [bors try](https://bors.tech/documentation/reference/). That way, you don't accidentally merge flaky code.

**Comment by matklad** [comment]
@notriddle actually I was hoping that the build passes by accident, because obviously my 100% correct changes could not have caused this unrelated build failure. I am so glad that bors forbids me to make such silly and unprofessional mistakes, which are nether the less very tempting after finishing a chunk of code :)

**Comment by matklad** [comment]
Spotted one more bit of logical duplication: `processFnsWithInherentPriority` and `associatedTypesTransitively` are doing a similar logic. I think what we probably need here is 

`RsTraitOrImpl.logicalMembers`, which for traits would be just direct memebers, and for impls direct members + members from the trait, which are not implemented in the impl directly. "transitiveness" (traits from bounds) should probably be handled separately, via `findImplsAndTraits`.

File: src/main/kotlin/org/rust/lang/core/resolve/NameResolution.kt [hunk 1]
```diff
@@ -435,27 +419,56 @@ private fun processFieldDeclarations(struct: RsFieldsOwner, subst: Substitution,
...
+    }
+    return lookup.derefTransitively(receiver).any { processAssociatedItems(lookup, it, VALUES, methodProcessor) }
+}
+
+private fun processAssociatedItems(lookup: ImplLookup, type: Ty, ns: Set<Namespace>, processor: RsResolveProcessor): Boolean {
+    val visitedInherent = mutableSetOf<String>()
+    fun processTraitOrImpl(traitOrImpl: BoundElement<RsTraitOrImpl>, inherent: Boolean): Boolean {
+        fun inherentProcessor(entry: BoundElement<RsNamedElement>): Boolean {
+            val name = entry.element.name ?: return false
+            if (inherent) visitedInherent.add(name)
```
Comment (vlad20012) [inline src/main/kotlin/org/rust/lang/core/resolve/NameResolution.kt]: Why name but not the element? The same name may be used in different scopes

**Pull request merged**